### ISSUE TITLE
Improve zfs snapshot restore error

### ISSUE THREAD

**Issue by raharper** [lxc/lxd#1815]
* Distribution: Ubuntu
 * Distribution version: 16.04 (Xenial)
 * The output of "lxc info" or if that fails:

```% lxc info```
apicompat: 0
auth: trusted
environment:
  addresses: []
  architectures:
  - x86_64
  - i686

 driver: lxc
 driverversion: 2.0.0.rc13
 kernel: Linux
 kernelarchitecture: x86_64
 kernelversion: 4.4.0-15-generic
 server: lxd
 serverpid: 7284
 serverversion: 2.0.0.rc6
 storage: zfs
 storageversion: "5"
config:
  core.trust_password: true
  storage.zfs_pool_name: lxd_zfs
public: false

lxc snapshots when using lxd do not warn that zfs backends only work with one backup.
However, when you restore, it warns you (when it's too late as you've already added another
snapshot).

# Steps to reproduce

 1. `lxc launch ubuntu t1`
 2. `lxc snapshot t1 first`
 3. `lxc snapshot t1 second`
 4. `lxc restore t1 first`

`% lxc restore t1 first`
error: ZFS only supports restoring state to the latest snapshot.

**Comment by stgraber** [comment]
You misunderstand the error, ZFS does support multiple snapshots, all the snapshots listed do exist and are perfectly usable, you just can't restore anything but last one.

So if you want to restore the before to last one, you have to remove the last one.

Alternatively, you can "copy" any snapshot you want into a new container with "lxc copy source-container/snapshot-name new-container"

**Comment by raharper** [comment]
It remains confusing highly.  Doubly so since I could see that I was allowed to create multiple snapshots.  Allowing one to view multiple snapshots via lxc info <container> and then having restore fail is highly confusing.  I do think it would be valuable to document, acknowledge the subtlety here.

The copy worked.  It would be nice to put that in the error message.  Something akin to:  We can only restore to the previous snapshot.  To use older snapshots, use lxc copy <machine>/<snapshot> <new-machine>

**Comment by stgraber** [comment]
Yeah, we can certainly tweak the error message a bit.

**Comment by nick4u** [comment]
This is not an lxd issue thou i think lxd doc should temporarily mention an important information about zfs dmks available at ubuntu 15.10

zfs-dkms and spl-dkms at at 15.10 is shipped with  0.6.4.2-0ubuntu1.1 version (http://packages.ubuntu.com/search?keywords=spl-dkms&searchon=names)

This version induces kernel panic if snapshots are accessed via hidden .zfs/snapshot folder
This feature is very useful if one would like to access snapshot data without restoring/cloning  dataset (for instance just access state of single file without rolling back entire fs)

xenial version has this fixed and Ubuntu 15.10 can have 
```
# dkms status
spl, 0.6.5.6, 4.2.0-34-generic, x86_64: installed
zfs, 0.6.5.6, 4.2.0-34-generic, x86_64: installed
```

via deb http://ppa.launchpad.net/zfs-native/stable/ubuntu wily main

 
![zfs kernel panic](https://cloud.githubusercontent.com/assets/5399008/14102665/482fa97e-f59c-11e5-8e2b-403c09349f9d.png)

**Comment by stgraber** [comment]
Can you file a bug in Ubuntu against the zfs-dkms package?

This should either be fixed or the feature should be turned off, but a kernel panic is a pretty wrong behavior :)

**Comment by stgraber** [comment]
https://launchpad.net/ubuntu/+source/zfs-dkms/+filebug

**Comment by nick4u** [comment]
1st. Above link does not work (This page does not exist, or you may not have permission to see it.)
2nd. This is old bug. I run onto it at debian a long time ago and get rid of it by using "deb http://archive.zfsonlinux.org/debian wheezy main" .

https://github.com/zfsonlinux/zfs/issues/3243
https://github.com/zfsonlinux/zfs/issues/3948
https://github.com/zfsonlinux/zfs/issues/3257
https://github.com/zfsonlinux/zfs/issues/3786

bug is fixed but for some reason ubuntu 15.10 still ships with old broken pkg 

Standard 15.10 instalation has: 
$ dkms status
spl, 0.6.4.2, 4.2.0-30-generic, x86_64: installed
spl, 0.6.4.2, 4.2.0-34-generic, x86_64: installed
zfs, 0.6.4.2, 4.2.0-30-generic, x86_64: installed
zfs, 0.6.4.2, 4.2.0-34-generic, x86_64: installed

I am not sure who to contact with to be honest, will try package maintainter from http://packages.ubuntu.com/wily-updates/spl-dkms

**Comment by HW-Siew** [comment]
Hi, I am wondering is there any ways we can 'store'/backup snapshot for a given container?  Let say i have 4 snapshot (snap1-4), and i wanted to restore to snap2. For now what i can do is to delete snap 3 & 4 to achieve that. But i might want to restore 3 or 4 later on? Is it possible to create a 'placeholder' container and park all snap under it?

Thanks for the advise.
